chrome/content/api/FPVS/implementation.js

```diff
@@ -0,0 +1,394 @@
+          windowListener.emit("dragdrop", win, "onDragLeave");
+          break;
+      }
+    }
+
+    add(callback) {
+      this.on("dragdrop", callback);
+      this.callbackCount++;
+
+      if (this.callbackCount == 1) {
+        ExtensionSupport.registerWindowListener(this.listenerId, {
+          chromeURLs: [
+            "chrome://messenger/content/messenger.xhtml"
+          ],
+          onLoadWindow: function (window) {
+            let folderPaneHeader = window.document.getElementById("folderPaneHeader");
+            folderPaneHeader.addEventListener("mouseout", windowListener, false);
+            folderPaneHeader.addEventListener("dragend", windowListener, false);
+            folderPaneHeader.addEventListener("dragenter", windowListener, false);
+
+            //left for testing when the new foldertree is enabled
+            //folderPaneHeader.addEventListener("drop", () => { log("foldpa drop"); }, false);
+            //folderPaneHeader.addEventListener("dragexit", () => { log("foldpa dragexit"); }, false);
+            //folderPaneHeader.addEventListener("dragover", () => { log("foldpa dragover"); }, false);
+            //folderPaneHeader.addEventListener("dragleave", () => { log("foldpa dragleave"); }, false);
+            //folderPaneHeader.addEventListener("mouseup", () => { log("foldpa mouseup"); }, false);
+            //folderPaneHeader.addEventListener("pointerup", () => { log("foldpa pointerup"); }, false);
+            //folderPaneHeader.addEventListener("dragover", () => { log("tree dragover"); }, false);
+            //folderTree.addEventListener("drop", () => { log("tree drop"); }, false);
+            //folderTree.addEventListener("dragend", () => { log("tree dragend"); }, false);
+
+            let folderTree = window.document.getElementById("folderTree");
+            folderTree.addEventListener("dragover", windowListener, false);
+            folderTree.addEventListener("dragleave", windowListener, false);
+          },
+        });
+
+        // Dragexit and dragdrop don't actually get sent when the user
+        // drops a message into a folder. This is arguably a bug in
+        // Thunderbird (see bz#674807). To work around it, I register a
+        // folder listener to detect when a move or copy is
+        // completed. This is gross, but appears to work well enough.
+        // By 2022, the bug is wontfix
+        // This needs to be registered only once, not per window!
+        MailServices.mfn.addListener(
+          this,
+          MailServices.mfn.msgsMoveCopyCompleted |
+          MailServices.mfn.folderMoveCopyCompleted
+        );
+      }
+    }
+
+    remove(callback) {
+      this.off("dragdrop", callback);
+      this.callbackCount--;
+
+      if (this.callbackCount == 0) {
+        log("Remove listener");
+        for (let window of ExtensionSupport.openWindows) {
+          if ([
+            "chrome://messenger/content/messenger.xhtml"
+          ].includes(window.location.href)) {
```

Comment:
Is it ensured that the `window.location.href` is always exactly the same URL as in the provided URL list? Or should optional parameters be stripped before comparison?

Reply:
Using url.origin to strip parameters now.